Factor out ImageSlice and ImageView from Image

pwinston opened the pull request:
# Description
Factor out two classes from `Image` in preparation for async loading for #1320.

## Type of change
- [x] Refactoring - no functional changes

# Explanation

`Image` is complicated and with async and async-multiscale it's going to get *way* more complicated. We need to start factoring out things that will need to change or grow.

Here we factor out two new classes `ImageSlice` and `ImageView`. Together they let us organize the 4 related images that `Image` stores, and makes it much harder to get into a broken state. Most importantly it gives us classes that can grow particularly `ImageSlice`.

The new classes might look like overkill to some people, but if you focus on the diff of `Image.py` it cleans up things a lot. I think we'll need to a lot more like this to prepare for growth.

| Before | After |
|---|---|
| _data | _data |
| _data_view | _slice.image.view |
| _data_raw | _slice.image.raw |
| _data_thumbnail | _slice.thumbnail.view |
| N/A | _slice.thumbnail.raw |

Problems before:
* `Image` class is messy, too much inline.
* The 4 _data values are confusingly named
    * `_data_raw` is not the raw version of `_data`!
    * Not clear which are the raw + view pairs
        * i.e. that `_data_thumbnail` is like `_data_view`, both are viewable and not raw.
    * The per-slice attributes are not grouped or distinguished enough from `_data`.
* Nothing enforces view will get updated when raw is updated.
* No where to put new Slice specific functionality.

Before set image and thumbnail raw:
```
    self._data_raw = np.clip(image, 0, 1)
    self._data_view = self._raw_to_displayed(self._data_raw)
    self._data_thumbnail = self._raw_to_displayed(	
        np.clip(thumbnail_source, 0, 1)	
    )	
```
After:
```
    self._slice.image.raw = np.clip(image, 0, 1)
    self._slice.thumbnail.raw = np.clip(thumbnail_source, 0, 1)
```

# Context and Comments

Pylint by default restricts you to at most 7 attributes per class. 7 is not random it's the [magical number](https://en.wikipedia.org/wiki/The_Magical_Number_Seven,_Plus_or_Minus_Two). More than 7 and it's hard to hold them in your head at the same time. The object becomes harder to understand, reason about, extend, test and debug. `Image` today has more than 70 attributes, public and private. You can see them in the debugger. A lot are through inheritance. 70 is way too many to understand or reason about. 

People think 7 is impossibly low but it's not that hard. When you add your 8th attribute you look at the those 8 and say is this really a list of 8 *totally unrelated* attributes? Is there really no connection between any of them? No pattern, no grouping? Or can you identify 2 or more than in fact do have some meaningful connection? If so combine those attributes into something new, a named tuple, dataclass, class, even a dict.

This sounds somewhat pointless at first, didn't you just move the attribute? How is that better? The benefit accrues over time because now there is a place to put functionality that deals with only those related attributes. Functionality that doesn't clutter the base object. Functionality that only has access to the small number of related attributes in its own class.

This new object then grows and then also splits, and so on recursively. The original object is now a tree of well factored objects with focused purposes. Each one taken alone is far easier understand, reason about, extend, test and debug than original kitchen-sink spaghetti object. Not to mention each one is potentially resable in ways the original class wasn't.

7 seems like a small number but consider if you only have 3 levels of hierarchy (a.b.c) then 7 gives you ~400 attributes (7 + 7^2 + 7^3). So your 3-level object could have much more state than `Image` yet no single object has more then 7 attributes. 

Some suggestions:
* Prefer small objects, a hard or soft limit of 7 attributes.
* Group attributes that are related: named similarly, updated together, derived from each other, etc.
* Prefer composition to inheritance, almost always.
* Prefer inheritance from small base classes, not big ones. Rather than inherit from `Image` have `Image` hold objects which have tiny inheritance hierarchies of their own.
* Prefer shallow inheritance hierarchies to deep ones.

Every rule and guideline can be broken in some cases if there's some strong reason to make an exception. However in general you need to constantly be pushing on these rules because code will constantly be growing in ways that violate them. This is almost entirely a problem related to **growth over time**. No one sits down and types up a class that has 50 attributes. They are added 1, 2 or 3 at a time by different people over a period of years. It's *almost always* easier to add attributes than to thoughtfully refactor things. Just like it's easier to throw something into the garage than to re-organize what's already there. And what can it hurt going from 17 to 18 attributes? What's one more?

# How has this been tested?
- [x] all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas

pwinston committed:
Factor out ImageSlice and ImageView

Previously:
    _data : the full data, all slices
    _data_view : viewable current slice
    _data_raw : raw current slice
    _data_thumbnail : viewable current thumbnail
    N/A             : raw current thumbnail

    Image class updates raw->view:
        self._data_raw = np.clip(image, 0, 1)
        self._data_view = self._raw_to_displayed(self._data_raw)

Now:
    _data : same as before
    _slice.image.view : viewable current slice
    _slice.image.raw : raw current slice
    _slice.thumbnail.view : viewable current thumbnail
    _slice.thumbnail.raw : raw current thumbnail

    ImageSlice class updates raw->view
        self._slice.image.raw = np.clip(image, 0, 1)

pwinston committed:
Move factored out method to creat a cleaner diff (hopefully)

pwinston committed:
Remove: not part of this refactoring

pwinston committed:
Extract 2 files for ImageSlice and ImageView

pwinston committed:
Better comments

pwinston committed:
Add comment

pwinston committed:
Comments, type annotations

pwinston committed:
Remove temp comment

d-v-b commented on napari/layers/image/_image_slice.py:
```diff
@@ -0,0 +1,43 @@
+    Example
+    -------
+        # Create with some default image.
+        slice = ImageSlice(default_image)
```
this shadows the python built-in `slice` object, maybe change to `imslice`?

pwinston replied:
I was like why didn't it warn me? But it was only in the comment. I changed it to `image_slice`, best to not shadow built-ins in examples.

pwinston committed:
Change to image_slice since slice is reserved (just in commment)

pwinston committed:
Keep Image._data_view and Image.data_raw for now

Some tests use Image.data_raw but tons of tests and non-tests
use Image._data_view. Maybe we can move away from it at some
point but for now we need it.

pwinston committed:
Remove line that we strongly believe was doing nothing

Nick on Zulip:
all the tests pass on master without it suggesting that it can be removed, i do have vague memories of why it was there, potentially related to passing "multiscale" labels and needing an initialization to be correct, but looking through the code now I'm really pretty confident it isn't needed anymore. At this point if removing it introduces a bug that isn't caught by our current tests I'd want to add a test and properly fix the bug instead of leaving this "hack" in so to speak

pwinston committed:
Fix labels layers: allow it to pass in a raw->view converter

This a more powerful and cleaner setup than having _raw_to_displayed
be a method of Image that gets overridden by Labels.

Since we pass in the converter, Labels or any other layer can pass
in any function they want, they don't have to just define one single
override. Also it remove the pointless "null" version of
_raw_to_displayed that Image used it have.

Minor points but still better.

pwinston committed:
Put _raw_to_displayed back in Image class

Tests forbid Image.__init__ from having any parameter that's
not returned by _get_state, so we can't pass in the converter
to Image.__init__.

Not a big deal this now works as before where the converter
is hardcoded based on the layer class. Less flexible but fine.

pwinston committed:
Remove now unused param to __init__

pwinston committed:
Fix doc string

sofroniewn reviewed (approved):
I really like this direction for the codebase. I find the chaining together of simple objects `layer._slice.image.view` much more natural and easy to remember that properties like `layer._data_view` or what might be `layer._slice_image_view`, and I really like how that leads to very transparent documentation with everything clearly having its own clear doc strings etc.

I've also pulled down the code and run through a battery of interactive demos and examples and everything worked fine. Our tests are also passing, and I love how this PR is fully backwards compatible!!! Great job there @pwinston.

I'd like @tlambert03 and @royerloic to review too before merge as it is a different way of organizing the codebase. If possible @pwinston and you're ok continuing to develop off of a branch of this PR for now we might also wait to merge until juan can take a look too.

I am also curious about where you see methods like `_set_view_slice` and `_update_thumbnail` living. That might help me get a better sense of how this continues to evolve (see https://github.com/napari/napari/pull/1343#issuecomment-639839349), but I get its also a work in progress.

kne42 commented on napari/layers/image/_image_slice.py:
```diff
@@ -0,0 +1,51 @@
+        """
+        Create an ImageSlice with some default viewable image.
+
+        Parameters
```
normally you don't have a docstring for `__init__` and put it in the class docstring (simply move the parameters section to there)

pwinston replied:
Just learned that in the other PR. Moved them for ImageSlice and ImageView here.

pwinston committed:
kn42 review: fix doc strings

pwinston committed:
Merge branch 'master' into refactor/image-slice

jni commented on napari/layers/image/_image_slice.py:
```diff
@@ -0,0 +1,51 @@
+
+        # Access the viewable images.
+        draw_image(image_slice.image.view)
+        draw_thumbnail(image_slice.thumbnail.view)
```
Can this be reformatted into doctest-compatible syntax? See examples [here](https://numpy.org/doc/stable/docs/howto_document.html), search for "Examples".

pwinston replied:
Updated to doctest format.

jni commented on napari/layers/image/_image_view.py:
```diff
@@ -0,0 +1,87 @@
+        image.view = view_image
+
+        # Set the raw image, it will compute the new viewable one.
+        image.raw = raw_image
```
Same here: use doctest format:

https://numpy.org/doc/stable/docs/howto_document.html

And the section name should be

```
Examples
--------
```

pwinston replied:
Updated to doctest format.

jni commented on napari/layers/image/_image_slice.py:
```diff
@@ -0,0 +1,51 @@
+    ----------
+    view_image : ArrayLike
+        The default image for the time and its thumbail.
+    image_converter : ImageConverter, optional
```
I think in the docstring (rather than the actual type annotation) I would prefer this to be `Callable, mapping array->array`, or something in that vein. When I read this, I see a crazy new object. Nothing in this docstring actually tells me, as a reader and longtime user of the scipy ecosystem, that `lambda x: x % 256` is an ImageConverter!

jni replied:
The description below should indeed explain what imageconverter should be, rather than how imageslice uses it, since readers typically read this docstring to figure out what they need to pass / can pass this constructor.

pwinston replied:
I agree in a way it would be nice to see the underlying type is `Callable[[ArrayLike], ArrayLike]` here. But I think that raises two questions:

1. Do we really want the type hint and doc string to purposely conflict? To me that looks like a mistake, and I see that mistake everywhere. Having the same type duplicated in two places where one is a comment is super error prone, against DRY/SPOT. They will inevitably get out of sync. But we are stuck with that format until they update things to suck out the type hints for docs.

2. If we keep them in sync and put `Callable[[ArrayLike], ArrayLike]` in both places are we just saying user defined types are not allowed? If not then when are they allowed? We have to spell that out I think, otherwise it's personal preference and thus hard to review.

In my experience using types without user defined types is the worst of all worlds and is a rookie mistake for people not used to typed languages. That said, I agree it can be overdone. Maybe some existing Python project's style guide discusses this?

Here's an example of where not using user defined types gets you, and a [basic discussion](https://realpython.com/python-type-checking/#type-aliases) of "type aliases" as Python calls user defined types:

<img width="599" alt="beazley-types" src="https://user-images.githubusercontent.com/4163446/84517070-bc2a8400-ac9c-11ea-8607-65414031471f.png">

tlambert03 replied:
> Do we really want the type hint and doc string to purposely conflict?

I've never seen the numpy docstrings as very useful for actually typing things.  There is no standard, and they're nearly impossible to parse.  So it's just there as a human readable string (and any type information is also just a human readable suggestion).  There _are_ ways to pull out [type hints for sphinx docs](https://pypi.org/project/sphinx-autodoc-typehints/)... so technically we don't need the type hints in the numpy docstrings at all.  But they are helpful as a suggestion if someone is looking at docstrings in an interactive session... I just don't see them as binding in any way, or something that can really be "synced" successfully with the (actual) type annotation.

I'm not sure what you mean by the second point.  If there's a standard type that we use in lots of places (like `ReaderFunction`, or `LayerData`), then we should definitely use type aliases like that.  But fwiw, i disagree with that twitter post.  I _like_ having those annotations like that.  It's immediately self-documenting, and I don't have to go searching around for what it means.  (And also, if you _really_ hate that... you can put the type hints into a separate `.pyi` file)

sofroniewn replied:
I can see arguments on both sides, and like the idea of finding a well respected style guide we can conform too (if such a thing exists!).

Given that this is something that we can very easily change and we'll be revisiting this part of the code shortly I'd like to merge now so as not hold up development (I'll be working on the world coordinates later today and so would like this in). This definitely doesn't mean though that I want to stop this convo or consider it settled.

pwinston replied:
I've heard this complaint many times about types, and about "too many" functions and methods. People feel like stuff is being "hidden from them" and they just want to see it all at once. Totally a guess but I feel like maybe 20% of programmers have this complaint. I could give you a list of names from over the years.

To me the twitter example above is beyond horrific. I wouldn't include code like that in any project I was working on unless my specific intent was to hinder the development process and sink that project. So that gives? How can reasonable people have 180 degree out of phase opinions on stuff like this? Over the years I've concluded two things:

1. People just think differently. Some people like to see all the parts spread all over the table and other people don't. There's no way or reason to try to change either person, they are just wired that way. FWIW I feel like the "smarter" people tend to like to see all the details at once, because they can make sense of it. And crucially they simply can't imagine how anyone would not like to see it that way.

2. It's possible to mentally or physically "undo" user defined types, or too many small functions or methods, with tools or by hand, and create the "all details at once view". But it's totally impossible to do the reverse. A good type system, or functional decomposition, reflects a **huge** amount of cognitive work, carefully breaking complex things down and coming up with meaningful names. So between the two approaches I think it's clear the one with more information has to win out, if any reasonable fraction of people like it that way.

Still this is one of those things where people will never agree, but there needs to be project style where people say "I personally would not write it like that, but it fits our style". I'm not asserting what the right style is for napari, but it needs to develop one or reviews will always go in circles as people push for their personal style.

Finally I can't remember where I first read this, but this is a really important anecdote. People like to say lines of code doesn't matter, but it matters a lot by order of magnitude. Consider a 200 line program vs. 2000, 20,000, 200,000, 2M, 20M, 200M. I've worked on 200 to 2M but not bigger.

What this person was saying, which I totally agree with, is if you are used to one size and you look at the next larger size you don't think "wow that's great code" you think "What the heck are they doing? That looks really clunky. Why are they making it so complicated? That's crazy I'd *never* write it like that. I could do it *much* more simply."

But if you expand your 2,000 line program into a 20,000 line program without adopting the styles of the next size up, everything grounds to halt and your codebase becomes an usable mess. Suddenly you are like "oh.... *now* I see why they did it that way". And the same with each "level jump" to the next larger size.

You see this "size discrepancy" all the time online because people argue about code without revealing what type of project they are working on. It's like you have 3 builders heatedly talking over lunch about building techniques, but one is building a dog house, one is building an big box store, and one is building a nuclear power plant. So they are all right but also all wrong.

pwinston replied:
I'll attempt a TLDR:
* Programmers sometimes have *very* different ideas about what style of code is easier or harder to understand or work with. Sometimes the exact opposite instincts.
* Projects need a good clear style so reviews can focus on that and not personal preferences.
* Some styles that work well in smaller codebases fall apart in larger ones.
* Corollary: some styles that work well in large codebases can be clunky in smaller codebases.

Also I was thinking that for me the case for using a type aliase or user defined type gets stronger the more times that a type appears in the code. Maybe there is a crossover-point where the benefits of "factoring out the duplication" outweigh the burden of "one more level of indirection".

tlambert03 replied:
> Also I was thinking that for me the case for using a type aliase or user defined type gets stronger the more times that a type appears in the code.

that's what I meant by pointing out `LayerData` and `ReaderFunction`, where we do just that.

jni replied:
> Also I was thinking that for me the case for using a type alias or user defined type gets stronger the more times that a type appears in the code.

That's exactly right, and I see a short future for ImageConverter, especially in a world in which NumPy is starting to develop its own annotation stubs, so an array -> array function will probably get its own name from there.

But, this is a bit of a side discussion on the main one, which was the NumPy style docstring type annotations. To answer your earlier questions, @pwiston,

- I don't think it's a huge problem if they diverge from the actual type annotation. One annotation is meant for humans, another for machines. Attempts to join the two modes have usually failed.
- the text description of the argument in the docstring certainly has *no* restrictions and that should clarify what we are talking about here.
- I actually completely disagree with Talley that these annotations are not useful — this is literally the first place I check when exploring a new API and if the object types are simple, the description is always sufficient for me to infer what I need to do. Contrast this with the standard library, which doesn't have this convention and I am often left wondering what exactly needs to go in that argument.

And of course, going back to the ImageConverter type alias — for this specific case, it is so simple that indeed I think `Callable[[np.ndarray], np.ndarray]` is preferable. It's not even that many more characters. And in fact, I think that type is too precise, since a callable mapping a dask array to a dask array would work here also. So it could just be `Callable`, or `Callable[[ArrayLike], ArrayLike]`.

> Maybe some existing Python project's style guide discusses this?

I think type hints are too new in Python for any style guides to have made pronouncements on this. But in Dave Beazley's example, I don't think type aliases would help. He would still hate the signature and it would be even less clear, unless, as discussed above, the type aliases are extremely common or totally self-explanatory. `ImageConverter` is neither of those, nor do I foresee it becoming so.

tlambert03 replied:
> I actually completely disagree with Talley that these annotations are not useful

? I feel like we're saying the exact same thing!  so I think you misunderstood my point.  I didn't say they are not useful "full stop"... I said they are not useful for actually typing things. (in a way that can be programmatically parsed), and I said they are there as human readable... exactly where a human might go first when exploring a new API, or in an interactive session.

jni replied:
Oops! Sorry for misunderstanding your point! Agreement is good! 😅

pwinston replied:
This PR is merged, I see you opened #1355 which is good. If the answer to when to use type aliases is "sometimes" we should flesh out in more detail what the guidelines are, when to use which style. Long term IDE's should let you toggle showing expanded types. You can see *some* expanded types on hover in vscode today. That seems to be heading towards the best of both worlds.

<img width="569" alt="ArrayLike" src="https://user-images.githubusercontent.com/4163446/84572894-20b41480-ad6b-11ea-9c33-6b508e4eaf48.png">

<img width="813" alt="LayerData" src="https://user-images.githubusercontent.com/4163446/84572898-2578c880-ad6b-11ea-8311-4e02da8c9858.png">

sofroniewn replied:
Those screenshots are pretty nice @pwinston. Maybe we continue this convo on #1355 and then if we do make changes have them be as part of a dedicated PR. I'd rather avoid mixing it in with #1354 where we'll have lots to discuss too.

jni replied:
My personal position on IDEs is that they are nice for people that have them, but policies should not depend on people using an IDE. Code should just be readable. Probably 90% of my interaction with code these days is reading it on GitHub PRs, so until GitHub itself becomes an IDE, plaintext interpretation is the lowest common denominator that we should target.

pwinston committed:
use doctest format in ImageView docstr

pwinston committed:
doctest for ImageView also

sofroniewn commented on napari/layers/image/_image_slice.py:
```diff
@@ -0,0 +1,54 @@
+        The main image for this slice.
+
+    thumbnail : ImageView
+        The smaller thumbnail image for this slice.
```
@pwinston just in case you were wondering about what this "thumbnail" actually is, it's not actually the thumbnail itself, but the sliced source data for the thumbnail, which for a normal image is just the image itself, but for a multiscale is the `_thumbnail_level` level of the pyramid.

Apologies that it is a bit weird, but hope that helps in case it was unclear. Probably when all is said and done we can figure out a better way to do this

pwinston replied:
Yeah I kind of suspected it was not really the thumbnail. One of the old variables was `thumbnail_source` and I probably shouldn't have shortened it. We should clear it up either just in docs or change the var names back. I'll look at in the next PR.

jni commented on napari/layers/image/_image_slice.py:
```diff
@@ -25,18 +25,21 @@ class ImageSlice:
+    Set raw image or thumbnail, viewable is computed.:
+
+    >> image_slice.image = raw_image
+    >> image_slice.thumbnail = raw_thumbnail
```
fwiw these should be triple `>>>` as the standard Python prompt.

Additionally, ideally all of our doctests will be executable and thus tested, which helps prevent them falling out of sync. When run, the are executed in the namespace of the module in which they live. So you'd want to add `  # doctest: +SKIP` at the end of lines that are meant to be explanatory rather than executed, which I think applies to the lines below?

pwinston replied:
I think I'll remove these `Examples` sections from the new PR for a few reasons:

1. I invented the idea of an **Example** section without knowing there was a standard that included doctest.

2. There's nothing to show here, there are no return values. It's pretty pointless.

3. No one is going to use these classes except where we do in the real system. No reason to pretend people will. Better to just explain how they work in regular docs.

I think an `Examples` section makes sense for some classes but not these.
